crates/virtio-queue/src/lib.rs

```diff
@@ -1582,4 +1685,154 @@ mod tests {
         q.state.next_avail = Wrapping(8);
         assert_eq!(q.enable_notification().unwrap(), false);
     }
+
+    #[test]
+    fn test_consume_chains_with_notif() {
+        let m = &GuestMemoryMmap::<()>::from_ranges(&[(GuestAddress(0), 0x10000)]).unwrap();
+        let vq = MockSplitQueue::new(m, 16);
+
+        let mut q = vq.create_queue(m);
+
+        // q is currently valid.
+        assert!(q.is_valid());
+
+        // The chains are (0, 1), (2, 3, 4), (5, 6), (7, 8), (9, 10, 11, 12).
+        for j in 0..13 {
+            let flags = match j {
+                1 | 4 | 6 | 8 | 12 => 0,
+                _ => VIRTQ_DESC_F_NEXT,
+            };
+
+            let desc = Descriptor::new((0x1000 * (j + 1)) as u64, 0x1000, flags, j + 1);
+            vq.desc_table().store(j, desc);
+        }
+
+        vq.avail().ring().ref_at(0).store(0);
+        vq.avail().ring().ref_at(1).store(2);
+        vq.avail().ring().ref_at(2).store(5);
+        vq.avail().ring().ref_at(3).store(7);
+        vq.avail().ring().ref_at(4).store(9);
+        // Let the device know it can consume chains with the index < 2.
+        vq.avail().idx().store(2);
+        // No descriptor chains are consumed at this point.
+        assert_eq!(q.next_avail(), 0);
+
+        let mut i = 0;
+
+        loop {
+            i += 1;
+            q.disable_notification().unwrap();
+
+            while let Some(_chain) = q.iter().unwrap().next() {
+                // Do something with the chain. Here the device would consume entries from the
+                // available ring, add an entry in the used ring and optionally notify the driver.
```

Comment:
I would also add a comment that for the purpose of this test what happens with the chain is not relevant, so we just ignore it as we just want to consume the entries in the queue.

Reply:
done